// ngrx-uml/dist/lib/services/create-actions-diagram.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateActionsDiagramService = void 0;
const __1 = require("../..");
const plant_uml_output_service_1 = require("../impl/outputs/plant-uml-output.service");
const actions_plant_diagram_renderer_factory_1 = require("../impl/renderers/actions-plant-diagram-renderer.factory");
const _1 = require("./");
class CreateActionsDiagramService {
    constructor(options) {
        this.options = options;
    }
    generateDiagram(filesPattern) {
        const plantUmlService = new plant_uml_output_service_1.PlantUmlOutputService({
            outDir: this.options.outDir || 'out',
            ext: this.options.imageFormat || 'png',
            clickableLinks: this.options.clickableLinks || false,
            saveWsd: this.options.saveWsd || false,
            generateDiagramsImages: this.options.generateImages || true
        });
        const generateService = new _1.GeneratorService([
            new __1.ActionConvertContextFactory,
            new __1.ActionReferenceConvertContextFactory,
        ], new actions_plant_diagram_renderer_factory_1.ActionsPlantDiagramRenderFactory().create(), [plantUmlService], this.options);
        return generateService.generate(filesPattern);
    }
}
exports.CreateActionsDiagramService = CreateActionsDiagramService;
