// ngrx-uml/dist/lib/core/converters/models/type-kind.enum.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeKind = void 0;
const typescript_1 = __importDefault(require("typescript"));
var TypeKind;
(function (TypeKind) {
    TypeKind[TypeKind["Unknown"] = 0] = "Unknown";
    TypeKind[TypeKind["TypeLiteral"] = 173] = "TypeLiteral";
    TypeKind[TypeKind["ExpressionWithTypeArguments"] = 216] = "ExpressionWithTypeArguments";
    TypeKind[TypeKind["TypeReference"] = 169] = "TypeReference";
    TypeKind[TypeKind["CallExpression"] = 196] = "CallExpression";
    TypeKind[TypeKind["VariableDeclaration"] = 242] = "VariableDeclaration";
    TypeKind[TypeKind["Identifier"] = 75] = "Identifier";
    TypeKind[TypeKind["PropertyAccessExpression"] = 194] = "PropertyAccessExpression";
    TypeKind[TypeKind["TypeArgument"] = 1001] = "TypeArgument";
    TypeKind[TypeKind["Action"] = 1002] = "Action";
    TypeKind[TypeKind["ActionReference"] = 1003] = "ActionReference";
    TypeKind[TypeKind["Reducer"] = 1004] = "Reducer";
    TypeKind[TypeKind["NgModule"] = 1005] = "NgModule";
    TypeKind[TypeKind["RegisteredReducer"] = 1006] = "RegisteredReducer";
})(TypeKind = exports.TypeKind || (exports.TypeKind = {}));
