// ngrx-uml/dist/lib/utils/serialize-symbols.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeClass = exports.serializeSignature = exports.serializeSymbol = void 0;
/** Serialize a symbol into a json object */
function serializeSymbol(symbol, checker) {
    return symbol.valueDeclaration && {
        name: symbol.getName(),
        // documentation: ts.displayPartsToString(symbol.getDocumentationComment(checker)),
        type: checker.typeToString(checker.getTypeOfSymbolAtLocation(symbol, symbol.valueDeclaration))
    };
}
exports.serializeSymbol = serializeSymbol;
/** Serialize a signature (call or construct) */
function serializeSignature(signature, checker) {
    return {
        parameters: signature.parameters.map(p => serializeSymbol(p, checker)),
        returnType: checker.typeToString(signature.getReturnType()),
    };
}
exports.serializeSignature = serializeSignature;
/** Serialize a class symbol information */
function serializeClass(symbol, checker) {
    const details = serializeSymbol(symbol, checker);
    if (symbol.valueDeclaration) {
        // Get the construct signatures
        const constructorType = checker.getTypeOfSymbolAtLocation(symbol, symbol.valueDeclaration);
        details.constructors = constructorType
            .getConstructSignatures()
            .map(s => serializeSignature(s, checker));
    }
    return details;
}
exports.serializeClass = serializeClass;
